### PR TITLE
[fx] Fix defining the ckpt functions inside forward function

### PR DESCRIPTION
**Pull request by Cypher30**
# What's New?

In this PR, I fix the small flaw in our previous codegen workflow that we actually define all the activation checkpoint functions inside our forward function, causing replicated defining of those functions. Here is an example of current codegen.

```python
# model definition
class MLP(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.linear1 = torch.nn.Linear(4, 4)
        self.linear2 = torch.nn.Linear(4, 4)

    def forward(self, x):
        return self.linear1(x), self.linear2(x)


class relu(torch.nn.Module):

    def __init__(self) -> None:
        super().__init__()
        self.relu = torch.nn.ReLU(inplace=True)

    def forward(self, x):
        return self.relu(x)


class MyModule(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.mlp1 = MLP()
        self.relu = relu()
        self.linear2 = torch.nn.Linear(4, 4)

    def forward(self, x):
        y1, y2 = checkpoint(self.mlp1, x)
        y3 = checkpoint(self.relu, x)

        def ckpt2(x):
            return F.relu(x, inplace=True)

        y4 = checkpoint(ckpt2, x)
        y4 = self.linear2(y4)
        return y1 + y2 + y3 + y4

# codegen forward code
class fx(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.mlp1 = torch.load(r'fx/mlp1.pt') # Module(   (linear1): Linear(in_features=4, out_features=4, bias=True)   (linear2): Linear(in_features=4, out_features=4, bias=True) )
        self.relu = torch.load(r'fx/relu.pt') # Module(   (relu): ReLU(inplace=True) )
        self.linear2 = Linear(in_features=4, out_features=4, bias=True)
        self.load_state_dict(torch.load(r'fx/state_dict.pt'))

    
    
    def checkpoint_0(self, x):
        mlp1_linear1 = self.mlp1.linear1(x)
        mlp1_linear2 = self.mlp1.linear2(x)
        return mlp1_linear1, mlp1_linear2
    def checkpoint_1(self, x):
        relu_relu = self.relu.relu(x)
        return relu_relu
    def checkpoint_2(self, x):
        relu = torch.nn.functional.relu(x, inplace = True);  x = None
        return relu
    def forward(self, x):
        import colossalai
        mlp1_linear1, mlp1_linear2 = colossalai.utils.activation_checkpoint.checkpoint(checkpoint_0, True, self, x, use_reentrant=True)
        relu_relu = colossalai.utils.activation_checkpoint.checkpoint(checkpoint_1, False, self, x, use_reentrant=False)
        relu = colossalai.utils.activation_checkpoint.checkpoint(checkpoint_2, False, self, x, use_reentrant=False)
        linear2 = self.linear2(relu);  relu = None
        add = mlp1_linear1 + mlp1_linear2;  mlp1_linear1 = mlp1_linear2 = None
        add_1 = add + relu_relu;  add = relu_relu = None
        add_2 = add_1 + linear2;  add_1 = linear2 = None
        return add_2
```

And I've run the test @super-dainiu built in our test, it works fine~

### CODE REVIEW

File: colossalai/fx/codegen/activation_checkpoint_codegen.py [hunk 1]
```diff
@@ -105,10 +105,10 @@ def _gen_ckpt_usage(label, activation_offload, input_vars, output_vars, use_reen
     """
     outputs = ', '.join(output_vars)
     inputs = ', '.join(input_vars)
-    return f'{outputs} = colossalai.utils.activation_checkpoint.checkpoint(checkpoint_{label}, {activation_offload}, {inputs}, use_reentrant={use_reentrant})'
+    return f'{outputs} = colossalai.utils.activation_checkpoint.checkpoint(checkpoint_{label}, {activation_offload}, self, {inputs}, use_reentrant={use_reentrant})'
```
Comment (FrankLeeeee) [inline colossalai/fx/codegen/activation_checkpoint_codegen.py:108]: Ideally this should be `checkpoint(self.checkpoint_{label}, ...)`, the current implementation does not bind the checkpoint method to the object.

Comment (Cypher30) [reply colossalai/fx/codegen/activation_checkpoint_codegen.py:108]: Ohhh yeah my bad

**Comment by FrankLeeeee** [comment]
An improved implementation will be pushed in a separate PR, so I will close this PR for now.
